### PR TITLE
feat: Add optional deposit and withdrawal fees

### PR DESCRIPTION
**Pull request by apollo-sturdy**

### CODE REVIEW

File: contracts/locked-astroport-vault/src/execute/basic_vault.rs [hunk 1]
```diff
@@ -79,6 +87,16 @@ pub fn execute_redeem(
     // Calculate claim amount and create msg to burn vault tokens
     let (burn_msg, claim_amount) = burn_vault_tokens(deps.branch(), &env, amount, &vt_denom)?;
 
+    // Deduct withdrawal fee if set
+    let claim_asset = Asset::cw20(base_token.clone(), claim_amount);
+    let fee_msgs = cfg.withdrawal_fee.fee_msgs_from_asset(claim_asset, &env)?;
+    let fee_amount = claim_amount * cfg.withdrawal_fee.fee_rate;
```
Comment (pacmanifold) [inline contracts/locked-astroport-vault/src/execute/basic_vault.rs:93]: how about having `fee_msgs_from_asset` return a tuple `(msgs: Vec<CosmosMsg>, assets_after_fee: AssetList)`?

File: contracts/locked-astroport-vault/tests/common/compound.rs [hunk 2]
```diff
@@ -35,7 +35,7 @@ pub fn test_compound_vault(
     // compound and check that the base token amount corresponding to the users
     // vault token balance has increased.
     let config = robot.query_config();
-    let treasury = config.treasury;
+    let treasury = &config.performance_fee.fee_recipients[0].0;
```
Comment (pacmanifold) [inline contracts/locked-astroport-vault/tests/common/compound.rs:38]: add to doc comment that first fee_recipient of performance_fee must be the treasury address

File: contracts/locked-astroport-vault/tests/test_deposit.rs [hunk 3]
```diff
@@ -77,3 +77,33 @@ fn can_only_deposit_when_despoits_enabled() {
...
+    let base_token_balance = robot.query_base_token_balance(user.address());
+    let deposit_amount = base_token_balance / Uint128::new(2);
+    let expected_amount =
+        deposit_amount * (Decimal::one() - fee_rate) * INITIAL_VAULT_TOKENS_PER_BASE_TOKEN;
+    robot
+        .deposit_cw20(deposit_amount, None, Unwrap::Ok, &user)
+        .assert_vault_token_balance_eq(user.address(), expected_amount)
+        .assert_total_vault_token_supply_eq(expected_amount)
+        .assert_total_vault_assets_eq(deposit_amount * (Decimal::one() - fee_rate));
+}
```
Comment (pacmanifold) [inline contracts/locked-astroport-vault/tests/test_deposit.rs:109]: add test for multiple fee recipients

File: contracts/locked-astroport-vault/tests/test_redeem.rs [hunk 4]
```diff
@@ -109,3 +115,47 @@ fn test_redeem_without_lockup() {
...
+            None,
+            &user,
+        )
+        .assert_base_token_balance_eq(
+            user.address(),
+            base_token_balance - deposit_amount * fee_rate,
+        )
+        .assert_vault_token_balance_eq(user.address(), Uint128::zero())
+        .assert_base_token_balance_eq(treasury.address(), deposit_amount * fee_rate);
+}
```
Comment (pacmanifold) [inline contracts/locked-astroport-vault/tests/test_redeem.rs:161]: add test with multiple fee recipients

File: contracts/locked-astroport-vault/tests/test_compound.rs [hunk 5]
```diff
@@ -12,12 +12,12 @@ pub mod common;
...
-#[test_case(VaultSetup::WstEth, Decimal::zero(); "Compound wsteth_eth vault, 0% fee.")]
-#[test_case(VaultSetup::AxlrNtrn, Decimal::zero(); "Compound axlr_ntrn vault, 0% fee.")]
-#[test_case(VaultSetup::WstEth, Decimal::percent(100); "Compound wsteth_eth vault, 100% fee.")]
-#[test_case(VaultSetup::AxlrNtrn, Decimal::percent(100); "Compound axlr_ntrn vault, 100% fee.")]
+#[test_case(VaultSetup::WstEth, Decimal::percent(5); "Compound wsteth_eth vault, 5% performance fee.")]
+#[test_case(VaultSetup::AxlrNtrn, Decimal::percent(5); "Compound axlr_ntrn vault, 5% performance fee.")]
+#[test_case(VaultSetup::WstEth, Decimal::zero(); "Compound wsteth_eth vault, 0% performance fee.")]
+#[test_case(VaultSetup::AxlrNtrn, Decimal::zero(); "Compound axlr_ntrn vault, 0% performance fee.")]
+#[test_case(VaultSetup::WstEth, Decimal::percent(100); "Compound wsteth_eth vault, 100% performance fee.")]
+#[test_case(VaultSetup::AxlrNtrn, Decimal::percent(100); "Compound axlr_ntrn vault, 100% performance fee.")]
```
Comment (pacmanifold) [inline contracts/locked-astroport-vault/tests/test_compound.rs:20]: test case with multiple fee recipients

File: contracts/locked-astroport-vault/src/state.rs [hunk 6]
```diff
@@ -44,6 +44,102 @@ pub fn claims() -> Claims<'static> {
...
+        Self {
+            fee_rate: value.fee_rate,
+            fee_recipients: value
+                .fee_recipients
+                .into_iter()
+                .map(|(addr, percentage)| (addr.to_string(), percentage))
+                .collect(),
+        }
+    }
+}
```
Comment (pacmanifold) [inline contracts/locked-astroport-vault/src/state.rs:141]: add some unit tests for `FeeConfig` functionality

Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/src/state.rs:150]: done

File: contracts/locked-astroport-vault/tests/test_compound.rs (hunk 5 above)
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/tests/test_compound.rs:20]: done

File: contracts/locked-astroport-vault/tests/test_redeem.rs (hunk 4 above)
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/tests/test_redeem.rs:161]: done

File: contracts/locked-astroport-vault/tests/test_deposit.rs (hunk 3 above)
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/tests/test_deposit.rs:199]: done

File: contracts/locked-astroport-vault/src/execute/basic_vault.rs [hunk 7]
```diff
@@ -87,29 +105,37 @@ pub fn execute_redeem(
...
-            claim_amount,
+            claim_amount_after_fee,
             cfg.lock_duration.after(&env.block),
         )?;
         let event = Event::new(UNLOCKING_POSITION_CREATED_EVENT_TYPE)
             .add_attribute(UNLOCKING_POSITION_ATTR_KEY, format!("{}", claim.id));
         Response::new().add_event(event)
     };
 
+    deps.api.debug(&format!("res: {:?}", res));
```
[the remark this answers is not in the document]
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/src/execute/basic_vault.rs]: done

File: contracts/locked-astroport-vault/src/execute/basic_vault.rs [hunk 8]
```diff
@@ -79,6 +87,16 @@ pub fn execute_redeem(
     // Calculate claim amount and create msg to burn vault tokens
     let (burn_msg, claim_amount) = burn_vault_tokens(deps.branch(), &env, amount, &vt_denom)?;
 
+    // Deduct withdrawal fee if set
+    let claim_asset = Asset::cw20(base_token.clone(), claim_amount);
+    let fee_msgs = cfg.withdrawal_fee.fee_msgs_from_asset(claim_asset, &env)?;
+    let fee_amount = claim_amount * cfg.withdrawal_fee.fee_rate;
+    let claim_amount_after_fee = claim_amount - fee_amount;
```
[the remark this answers is not in the document]
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/src/execute/basic_vault.rs]: done

File: contracts/locked-astroport-vault/src/execute/basic_vault.rs [hunk 9]
```diff
@@ -33,16 +33,22 @@ pub fn execute_deposit(
...
-        Asset::cw20(base_token, amount).transfer_from_msg(depositor, &env.contract.address)?,
-    );
+    let deposit_asset = Asset::cw20(base_token, amount);
+    let transfer_from_res = Response::new()
+        .add_message(deposit_asset.transfer_from_msg(depositor, &env.contract.address)?);
+
+    // Take deposit fee if set
+    let fee_msgs = cfg.deposit_fee.fee_msgs_from_asset(deposit_asset, &env)?;
+    let fee_amount = amount * cfg.deposit_fee.fee_rate;
+    let amount_after_fee = amount - fee_amount;
```
[the remark this answers is not in the document]
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/src/execute/basic_vault.rs]: done

File: contracts/locked-astroport-vault/tests/common/compound.rs (hunk 2 above)
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/tests/common/compound.rs:38]: I mean this is just to test the balance. Could also call it fee recipient.

File: contracts/locked-astroport-vault/src/migrations.rs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    /// The fee that is taken on rewards accrued
+    pub performance_fee: Decimal,
+    /// The router contract address
+    pub router: CwDexRouterBase<Addr>,
+    /// The asset to which we should swap reward_assets into before providing
+    /// liquidity. Should be one of the assets in the pool.
+    pub reward_liquidation_target: AssetInfoBase<Addr>,
+    /// Helper for providing liquidity with unbalanced assets.
+    pub liquidity_helper: LiquidityHelperBase<Addr>,
+}
```
Comment (pacmanifold) [inline contracts/locked-astroport-vault/src/migrations.rs:34]: could this not be imported from old crate?

File: contracts/locked-astroport-vault/src/state.rs [hunk 11]
```diff
@@ -188,3 +320,128 @@ pub struct StateResponse {
...
+    fn fee_msgs_from_asset_works() {
+        let env = mock_env();
+
+        let fee_config = super::FeeConfig {
+            fee_rate: Decimal::percent(1),
+            fee_recipients: vec![(Addr::unchecked("addr1"), Decimal::percent(100))],
+        };
+        let asset = Asset::new(AssetInfo::native("uusdc"), 100u128);
+        let (msgs, asset_after_fee) = fee_config.fee_msgs_from_asset(asset, &env).unwrap();
+        assert_eq!(msgs.len(), 1);
```
Comment (pacmanifold) [inline contracts/locked-astroport-vault/src/state.rs:391]: assert the message is correct?

File: contracts/locked-astroport-vault/src/state.rs [hunk 12]
```diff
@@ -188,3 +320,128 @@ pub struct StateResponse {
...
+            Asset::new(AssetInfo::native("uusdc"), 100u128),
+            Asset::new(AssetInfo::native("uatom"), 200u128),
+        ]
+        .into();
+        let (msgs, assets_after_fee) = fee_config.fee_msgs_from_assets(&assets, &env).unwrap();
+        println!("{:?}", msgs);
+        assert_eq!(msgs.len(), 2);
+        assert_eq!(assets_after_fee.to_vec()[0].amount, Uint128::new(99));
+        assert_eq!(assets_after_fee.to_vec()[1].amount, Uint128::new(198));
+    }
```
Comment (pacmanifold) [inline contracts/locked-astroport-vault/src/state.rs:427]: add unit test case for this case but with multiple recipients

File: contracts/locked-astroport-vault/src/state.rs [hunk 13]
```diff
@@ -188,3 +320,128 @@ pub struct StateResponse {
...
+            fee_recipients: vec![(Addr::unchecked("addr1"), Decimal::percent(100))],
+        };
+        let assets = vec![
+            Asset::new(AssetInfo::native("uusdc"), 100u128),
+            Asset::new(AssetInfo::native("uatom"), 200u128),
+        ]
+        .into();
+        let (msgs, assets_after_fee) = fee_config.fee_msgs_from_assets(&assets, &env).unwrap();
+        println!("{:?}", msgs);
+        assert_eq!(msgs.len(), 2);
```
Comment (pacmanifold) [inline contracts/locked-astroport-vault/src/state.rs:424]: assert messages are correct

File: contracts/locked-astroport-vault/tests/test_redeem.rs (hunk 4 above)
Comment (pacmanifold) [reply contracts/locked-astroport-vault/tests/test_redeem.rs:161]: can't find test for withdrawal fee with multiple recipients

File: Makefile.toml [hunk 14]
```diff
@@ -123,7 +123,7 @@ command = "cargo"
 args = [
     "build",
     "-p",
-    "locked-astroport-vault",
+    "locked-astroport-vault@0.3.0",
```
Comment (pacmanifold) [inline Makefile.toml:126]: why is this needed?

**Review by pacmanifold — COMMENTED** [review]
some more comments

File: contracts/locked-astroport-vault/src/migrations.rs (hunk 10 above)
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/src/migrations.rs]: done

File: Makefile.toml (hunk 14 above)
Comment (apollo-sturdy) [reply Makefile.toml:126]: because tree now includes two versions of the crate

File: contracts/locked-astroport-vault/src/execute/basic_vault.rs [hunk 15]
```diff
@@ -72,44 +78,61 @@ pub fn execute_redeem(
...
-        Response::new().add_event(event)
+        res.add_event(event)
     };
 
     let event = Event::new("apollo/vaults/execute_redeem")
         .add_attribute("is_force_redeem", format!("{}", force_redeem))
         .add_attribute("vault_tokens_redeemed", amount)
-        .add_attribute("lp_tokens_claimed", claim_amount);
+        .add_attribute("lp_tokens_claimed", claim_amount)
+        .add_attribute("withdrawal_fee", fee_amount);
```
[the remark this answers is not in the document]
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/src/execute/basic_vault.rs]: done

File: contracts/locked-astroport-vault/src/state.rs (hunk 11 above)
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/src/state.rs:391]: done

File: contracts/locked-astroport-vault/src/state.rs (hunk 13 above)
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/src/state.rs:431]: done

File: contracts/locked-astroport-vault/src/state.rs (hunk 12 above)
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/src/state.rs]: done

File: contracts/locked-astroport-vault/tests/test_redeem.rs (hunk 4 above)
Comment (apollo-sturdy) [reply contracts/locked-astroport-vault/tests/test_redeem.rs:161]: done

**Review by pacmanifold — APPROVED** [review]
lgtm
